clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -267,50 +180,82 @@ public T get() throws InterruptedException, ExecutionException {
     @Override
     public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException,
             TimeoutException {
-        SingleWaiter<T> waiter = new SingleWaiter<>();
-        addWaiter(waiter);
-        return waiter.await(timeout, unit);
+        try {
+            return completableFuture.get(timeout, unit);
+        } catch (ExecutionException e) {
+            maybeThrowCancellationException(e.getCause());
+            throw e;
+        }
     }
 
     /**
      * Returns the result value (or throws any encountered exception) if completed, else returns
      * the given valueIfAbsent.
      */
     @Override
-    public synchronized T getNow(T valueIfAbsent) throws InterruptedException, ExecutionException {
-        if (exception != null)
-            wrapAndThrow(exception);
-        if (done)
-            return value;
-        return valueIfAbsent;
+    public synchronized T getNow(T valueIfAbsent) throws ExecutionException {
```

Comment:
Does it need to remove `InterruptedException` from other get methods also? For another, why we don't remove `InterruptedException` from `KafkaFuture#getNow`?

Reply:
> Does it need to remove InterruptedException from other get methods also?

`get()` and `get(long, TimeUnit)` can still throw `InterruptedException` due to the blocking behaviour (the `completableFuture.get` that we delegate to throws `InterruptedException`).

> For another, why we don't remove InterruptedException from KafkaFuture#getNow?

Good spot!  `KafkaFuture#getNow` should never have declared that and it's annoying for callers to have to deal with it. But this wasn't described in the KIP and wouldn't be a source compatible change (existing code with a `catch (InterruptedException)` would get a compile error). 

I'd be happy to remove it and notify the VOTE thread, WDYT @kkonstantine @dajac ?

Reply:
Good question. AK 3.0 is a good opportunity to do such breaking changes so I would be in favour of doing it. Let's see what other think.

Reply:
> But this wasn't described in the KIP and wouldn't be a source compatible change (existing code with a catch (InterruptedException)

As it can cause compatible change, we should deprecate it and then add a new method (for example: get(T value)) to replace it. This can be discussed in another issue :)

Reply:
I wouldn't call it `get` because it has different semantics to the existing get methods, so I think it should have a distinct name. And I can't think of a succinct alternative to `getNow`. Something like `getOrDefault` doesn't convey they key difference from `get` (the lack of blocking). `getImmediately` is just a bit verbose. 

The API of KafkaFuture has always been a subset of `CompletableFuture`. Introducing a new method would break that. While it would be discoverable while the deprecated method remained to point the way it wouldn't be as discoverable once the old `getNow` was removed.

All deprecation really does is let the programmer defer fixing the problem. But it's most likely not a huge problem to fix (i.e. they've likely caught the exception close to the method call, rather than defining many layers of API with `throws InterruptedException`).

A new method also means their code is then not binary compatible with older client versions. 

So on balance I personally favour just removing the `throws InterruptedException` in 3.0 and letting the programmer clean up the code sooner rather than later. 

There is always the "do nothing" alternative.

Reply:
"do nothing" is probably the right thing here.

Reply:
OK, I've reverted it.

Reply:
> A new method also means their code is then not binary compatible with older client versions.

yep, the BC could be broken. What I really care is the source compatibility. It seems to me a public API should be source compatible to next major release :)

This patch is good to go and we need more time (rather than vespene gas ... StarCraft is my favor game :) ) to reach consensus.